DOCS-6008 connections (#170)
mulesoft/docs-mule-sdk: modules/ROOT/pages/connections.adoc

endif::[]
:keywords: mule, sdk, connection, connectivity, management, pooling, cached, provider, connection-provider

Connected modules (connectors) are the most common type of module. They provide a connection between a Mule flow and an external resource.
Configure these connections with the `ConnectionProvider` in a connectivity framework to create, disconnect, and describe the connection for Mule runtime engine to handle.

> For more information on how to use connections, refer to <<use-connections, this section>>.

Reviewer: ```suggestion
For more information about how to use connections, refer to <<use-connections>>.
```

Reviewer: (Name the section, rather than using "this section".)

Reviewer: Kim would say use "about" not "on" :)

== Declaring a Connection Provider

To declare a new `ConnectionProvider` you should create a new class that implements the `ConnectionProvider` interface. The `ConnectionProvider` interface is a typed interface that receives one type argument that indicates the class of the instances that the `ConnectionProvider` handles.

> This example declares a `myConnectionProvider` class that implements the `ConnectionProvider` interface:

Reviewer 2: by naming convention, class name should start with capital letter, so I'd replace _myConnectionProvider_ with _MyConnectionProvider_

[source,java,linenums]
----

> public class myConnectionProvider implements ConnectionProvider<myConnection> {

Reviewer 2: same here:

public class MyConnectionProvider implements ConnectionProvider<MyConnection> {

  @Parameter
  private String password;

  public myConnection connect() throws ConnectionException {

>     return new myConnection(password);

Reviewer 2: replace _myConnection_ with _MyConnection_

  }

>   public void disconnect(myConnection connection) {

Reviewer 2: replace with MyConnection

    connection.invalidate();
  }

>   public ConnectionValidationResult validate(myConnection connection) {

Reviewer 2: replace with MyConnection

[source,java,linenums]
----
@Extension(name="xyz")
@Operations(Operations.class)

> @ConnectionProviders(myConnectionProvider.class)

Reviewer 2: replace by MyConnectionProvider

> public class myModule {

Reviewer 2: replace with MyModule

Implementing the `ConnectionProvider` interface means no strategy.

[[use-connections]]
== Using Connections

Using connections in Operations and Sources is quite different, see:

* <<operations#_using_connections, Using Connections in Operations>>

> * <<sources-config-connection#, Using Connections in Sources>>

Reviewer: What is the # doing in this case? (I've never seen that construction.)
